dagutils/traversal.go

```diff
@@ -0,0 +1,507 @@
+// would just create a proxy for every move call, e.g., `dt.pos.down()`.
+//
+// TODO: Consider adding methods that would retrieve an attribute indexed
+// by the current `level` to avoid including much indexing like
+// `dt.path[dt.level]` in the code. Maybe another more strong refactoring
+// that would allow to think of the current node in the `path` without
+// concerning at what `level` it's in.
+type Traversal struct {
+
+	// Each member of the slice is the parent of the following member, from
+	// the root to the current node, up until the position marked by `level`. The
+	// slice may contain more elements past that point but they should be
+	// ignored, every time the `level` increases the corresponding child node
+	// is inserted (the slice is not truncated to leverage the already allocated
+	// space). It should *always* have a length bigger than zero with the root
+	// of the DAG at the first position (empty DAGs are not valid).
+	path []ipld.Node
+
+	// This slice has the index of the child each parent in `path` is pointing
+	// to. Its current valid length is determined by `level`. The index in the
+	// parent can be set past all of its child nodes (having a value equal to
+	// `len(Links())`) to signal it has visited (or skipped) all of them. A
+	// leaf node with no children and its index in zero would also comply with
+	// this format.
+	childIndex []uint
+	// TODO: Revisit name, `childPosition`? Do not use the *link* term, put
+	// emphasis in the child (the *what*, not the *how* we get it).
+
+	// Depth of the node of the current position. It grows downwards, root is
+	// level 0, its child level 1, and so on. It controls the effective length
+	// of `path` and `childIndex`.
+	//
+	// A level of -1 signals the start case of a new `Traversal` that hasn't
+	// moved yet (or its position has been reset, see `ResetPosition`).
+	// Although this state is an invalid index to the slices, it allows to
+	// centralize all the visit calls in the `down` move (starting at zero
+	// would require a special visit case inside every traversal operation like
+	// `Iterate()` and `Search`). This value should never be returned to after
+	// the first `down` movement, moving up from the root should always return
+	// `ErrUpOnRoot`.
+	level int
+
+	// Method called each time a node is arrived upon in a traversal (through
+	// the `down()` movement, that is, when visiting it for the first time,
+	// not when going back up). It is the main API to implement functionality
+	// (like read and seek) on top of the `Traversal` structure.
+	//
+	// Its argument is the current `node` being visited. Any error it returns
+	// (apart from the internal `errDownStopIteration`) will be forwarded to
+	// the caller of the traversal operation (stopping it).
+	//
+	// Any of the exported methods of this API should be allowed to be called
+	// from within this method.
+	VisitHandler func(node ipld.Node) error
```

Comment:
I might call this a `Visitor`. In fact, I think that’s generally the common term I’ve seen for it.

Reply:
If we want to re-use this across the go-ipfs codebase the Visit function should:

* Receive the `level` of the visited node
* Receive the parent(s) of the visited node

We probably want to provide a couple default Visit functions to wrap some commonly used logic:

* One implementing full recursive traversal of the dag, pruning already visited branches
* One implementing depth limited traversal, pruning already limited branches (if they are not going to be explored deeper).